### PR TITLE
Integrate WhisperKit library to Hub

### PR DESCRIPTION
**Pull request by atiorh**
cc: @Vaibhavs10 @pcuenca 

Goal is to keep an accurate count of the downloads for Core ML models supported by WhisperKit which match the pattern `**/model.mil`. Example library impacted by this: https://huggingface.co/argmaxinc/whisperkit-coreml

### CODE REVIEW

File: packages/tasks/src/model-libraries.ts
Comment (Wauplin) [inline packages/tasks/src/model-libraries.ts:380]:
```suggestion
			wildcard: { path: "*/model.mil" },
```

I expect this to be the correct way for ES wildcards

**Review by Wauplin — COMMENTED** [review]
Thanks for the PR @atiorh :) I saw you're already in discussion with @Vaibhavs10 about this. Just to let you know, you can also define the `snippets` fields to show users how to use the weights with your library.

Also, I noticed all models are bundled in a single repo [argmaxinc/whisperkit-coreml](https://huggingface.co/argmaxinc/whisperkit-coreml). Usually we tend to prefer to split them with 1 repo == 1 model. This way, each model can have their own modelcard + download count would be individualized. Users would be able to search them on this page: https://huggingface.co/models?other=whisperkit (with `library=whisperkit` once this PR is merged).

FYI, it is possible to programmatically manage repos on the Hub using the [`huggingface_hub`](https://github.com/huggingface/huggingface_hub) Python library so splitting the repo shouldn't be too difficult. You can then manage updates programmatically as well (to update versions for example). Please let me know if you want some assistance in this direction :)

File: packages/tasks/src/model-libraries.ts
Comment (atiorh) [reply packages/tasks/src/model-libraries.ts:380]:
```shell
ati@atis-MBP-2 whisperkit-coreml % ls -al **/model.mil | wc -l
      42
ati@atis-MBP-2 whisperkit-coreml % ls -al */model.mil | wc -l 
zsh: no matches found: */model.mil
       0
```

**Comment by atiorh** [comment]
> Thanks for the PR @atiorh :) I saw you're already in discussion with @Vaibhavs10 about this. Just to let you know, you can also define the `snippets` fields to show users how to use the weights with your library.
> 
> Also, I noticed all models are bundled in a single repo [argmaxinc/whisperkit-coreml](https://huggingface.co/argmaxinc/whisperkit-coreml). Usually we tend to prefer to split them with 1 repo == 1 model. This way, each model can have their own modelcard + download count would be individualized. Users would be able to search them on this page: https://huggingface.co/models?other=whisperkit (with `library=whisperkit` once this PR is merged).
> 
> FYI, it is possible to programmatically manage repos on the Hub using the [`huggingface_hub`](https://github.com/huggingface/huggingface_hub) Python library so splitting the repo should be too difficult. You can then manage updates programmatically as well (to update versions for example). Please let me know if you want some assistance in this direction :)

Thanks for the response @Wauplin! Our design for model management is based on having a single repo for all models at the moment but thank you for the suggestion. @Vaibhavs10  suggested the same and we will actively consider it. Other than that, do you see any blockers to merge? Looking forward to getting this in :)

**Comment by atiorh** [comment]
Not sure how but I accidentally requested a review from @julien-c  🤔

File: packages/tasks/src/model-libraries.ts
Comment (Wauplin) [reply packages/tasks/src/model-libraries.ts:380]: The wildcard pattern passed here correspond to an Elasticsearch string filter. This means the whole filepath is considered as a string, which is not the same as in `ls` command.

**Comment by coyotte508** [comment]
cc @Kakulukian for wildcard (you had a better way?)

**Comment by Kakulukian** [comment]
> cc @Kakulukian for wildcard (you had a better way?)

Not for the moment we are in the process of converting them but it's not yet done (query looks ok)

**Review by Vaibhavs10 — APPROVED** [review]
Sorry for the late response (I was out sick): LGTM - pending the wildcard query change suggested by @Wauplin 🙏)

In general, for later releases let's explore having one model checkpoint per repo. We can help from swift side (cc: @pcuenca knows best) if it helps facilitate that.

**Review by Wauplin — APPROVED** [review]
Great, then let's get this merged as it is now! :tada:

**Comment by atiorh** [comment]
Great! Just pushed the wildcard change. @Wauplin  please do the honors :)

**Comment by atiorh** [comment]
I think a maintainer needs to run the checks 🤔

File: packages/tasks/src/model-libraries.ts
Comment (osanseviero) [inline packages/tasks/src/model-libraries.ts:375]: This broke the filter, it's missing a closing one

```suggestion
	},
	"whisperkit": {
```

**Comment by atiorh** [comment]
Thanks for fixing the test @osanseviero 🙌  I tried fixing the Lint put I have no idea what I am doing in JS land..

**Comment by julien-c** [comment]
come to the JS land @atiorh it's nice!

File: packages/tasks/src/model-libraries.ts
Comment (pcuenca) [inline packages/tasks/src/model-libraries.ts:381]: Question: why are we tracking `model.mil`?

- WhisperKit repos, even if segregated, contain 3 Core ML models each.
- The way download stats are updated in other libraries is by pinging the repo for a known file (like `config.json`) to retrieve the last revision and determine whether the cache should be invalidated. Every ping (a `HEAD` request, usually) is a download. This is not yet supported in `swift-transformers`, but should come soon.

Tracking config.json would be compatible with the current download mechanism (snapshot download) and with the upcoming revision ping.

Otherwise this looks great to me!

File: packages/tasks/src/model-libraries.ts
Comment (pcuenca) [inline packages/tasks/src/model-libraries.ts:376]:
```suggestion
	whisperkit: {
```

`prettier` insists on not using `"` unless necessary.

File: packages/tasks/src/model-libraries.ts
Comment (atiorh) [reply packages/tasks/src/model-libraries.ts:381]: We were actually pinging `config.json` but it wasn't counting the stats probably because of the `swift-transformers` future work you mentioned? Happy to switch back to `config.json` when we can verify that route works in the future 🙌

Comment (pcuenca) [reply packages/tasks/src/model-libraries.ts:381]: You were pinging `config.json` inside subfolders IIRC, but only the root `/config.json` is counted by default. `*/config.json` would work for the same reason `*/model.mil` does (except once per repo instead of 3).

The future work is about counting every time the models are loaded, even if cached.

Comment (coyotte508) [reply packages/tasks/src/model-libraries.ts:381]: @pcuenca do you want to switch to `*/config.json`?

Comment (pcuenca) [reply packages/tasks/src/model-libraries.ts:381]: No, let's wait for the library owners to decide and open a new PR if needed. Thank you @coyotte508 😄

Comment (atiorh) [reply packages/tasks/src/model-libraries.ts:381]: > The future work is about counting every time the models are loaded, even if cached.
Ah now I understand! BTW, @coyotte508, would this rule apply to existing repos or only new ones? e.g. https://huggingface.co/argmaxinc/whisperkit-coreml is still showing the previous count. I added `library_name: whisperkit` in the model card in addition to the tag `-whisperkit`

Comment (Wauplin) [reply packages/tasks/src/model-libraries.ts:381]: @atiorh the counting rule will apply to all repos but it usually needs a few days to update correctly everywhere. After this transition period, you should see the correct numbers.

Comment (atiorh) [reply packages/tasks/src/model-libraries.ts:381]: Thanks for confirming!

Comment (osanseviero) [reply packages/tasks/src/model-libraries.ts:381]: >hmm the new rule will only be used to compute "new" downloads

No, as far as I know. Our elastic script recomputes the downloads of the last 30 days using the filter defined here. When we added Lora filters for diffusion models it gave us some nice stats

Comment (coyotte508) [reply packages/tasks/src/model-libraries.ts:381]: It's up on the hub by the way, JFI

**Pull request merged**
